src/jackdaw/streams/xform.clj

```diff
@@ -0,0 +1,67 @@
+(ns jackdaw.streams.xform
+  "Helper functions for working with transducers."
+  (:gen-class)
+  (:require [jackdaw.serdes :as js]
+            [jackdaw.streams :as j])
+  (:import org.apache.kafka.streams.kstream.Transformer
+           org.apache.kafka.streams.KeyValue
+           [org.apache.kafka.streams.state KeyValueStore Stores]
+           org.apache.kafka.streams.StreamsBuilder))
+
+(defn fake-kv-store
+  "Creates an instance of org.apache.kafka.streams.state.KeyValueStore
+  with overrides for get and put."
+  [init]
+  (let [store (volatile! init)]
+    (reify KeyValueStore
+      (get [_ k]
+        (clojure.core/get @store k))
+
+      (put [_ k v]
+        (vswap! store assoc k v)))))
+
+(defn kv-store-swap-fn
+  "Takes an instance of KeyValueStore, a function f, and map m, and
+  updates the store in a manner similar to `clojure.core/swap!`."
+  [^KeyValueStore store f m]
+  (let [ks (keys (f {} m))
+        prev (reduce (fn [p k]
+                       (assoc p k (.get store k)))
+                     {}
+                     ks)
+        next (f prev m)]
+    (doall (map (fn [[k v]] (.put store k v)) next))
+    next))
+
+(defn add-state-store!
+  [builder]
+  "Takes a builder and adds a state store."
+  (doto ^StreamsBuilder (j/streams-builder* builder)
+    (.addStateStore (Stores/keyValueStoreBuilder
+                     (Stores/persistentKeyValueStore "transducer")
+                     (js/edn-serde)
+                     (js/edn-serde))))
+  builder)
+
+(defn transformer
+  "Takes a transducer and creates an instance of
+  org.apache.kafka.streams.kstream.Transformer with overrides for
+  init, transform, and close."
+  [xf]
+  (let [ctx (atom nil)]
+    (reify
+      Transformer
+      (init [_ context]
+        (reset! ctx context))
+      (transform [_ k v]
+        (let [^KeyValueStore store (.getStateStore @ctx "transducer")
+              v (first (into [] (xf store) [[k v]]))]
+          (KeyValue/pair k v)))
+      (close [_]))))
+
+(defn transduce-kstream
+  [kstream xf]
+  "Takes a kstream and xf and transduces the stream."
+  (-> kstream
+      (j/transform (fn [] (transformer xf)) ["transducer"])
+      (j/flat-map (fn [[_ v]] v))))
```

Comment:
Why is the flat-map necessary here? What if someone wants to keep the keys?

Reply:
Flat map corresponds to the reducing function in `transduce`. We always `transduce` with `concat`. This allows you to take a record and transform it stateful into zero or more records. Without this step, you can't implement either of the examples.

Reply:
I think you could achieve the same functionality using the `forward` and `commit` methods on `ctx` (which is a [ProcessorContext](https://kafka.apache.org/10/javadoc/org/apache/kafka/streams/processor/ProcessorContext.html)). There's an example [here](https://github.com/bobby/kafka-streams-clojure/blob/master/src/kafka_streams_clojure/api.clj#L32). Not saying that way's better, just that there is an alternative way.

Reply:
I was going to make the same point!  I used ProcessorContext in the example @DaveWM linked because it felt closer to a transducible context than anything in the high-level KStreams API.  ProcessorContext also gives a bit more fine-grained control over state stores.